// @midwayjs/decorator/dist/annotation/objectDef.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Autowire = exports.Scope = exports.Destroy = exports.Init = exports.Async = void 0;
const __1 = require("../");
const debug = require('util').debuglog('decorator:context:obj_def');
function Async() {
    return function (target) {
        debug(`set [async] property in [${target.name}]`);
        return (0, __1.saveObjectDefProps)(target, { isAsync: true });
    };
}
exports.Async = Async;
function Init() {
    return function (target, propertyKey) {
        debug(`set [init] property in [${target.constructor.name}]`);
        return (0, __1.saveObjectDefProps)(target.constructor, { initMethod: propertyKey });
    };
}
exports.Init = Init;
function Destroy() {
    return function (target, propertyKey) {
        debug(`set [destroy] property in [${target.constructor.name}]`);
        return (0, __1.saveObjectDefProps)(target.constructor, {
            destroyMethod: propertyKey,
        });
    };
}
exports.Destroy = Destroy;
function Scope(scope = __1.ScopeEnum.Singleton) {
    return function (target) {
        debug(`set [scope] property in [${target.name}]`);
        return (0, __1.saveObjectDefProps)(target, { scope });
    };
}
exports.Scope = Scope;
function Autowire(isAutowire = true) {
    return function (target) {
        debug(`set [autowire] property in [${target.name}]`);
        return (0, __1.saveObjectDefProps)(target, { isAutowire });
    };
}
exports.Autowire = Autowire;
