// midway-web/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
// 兼容老代码
var decorator_1 = require("@midwayjs/decorator");
exports.configuration = decorator_1.Configuration;
exports.provide = decorator_1.Provide;
exports.inject = decorator_1.Inject;
exports.async = decorator_1.Async;
exports.init = decorator_1.Init;
exports.destroy = decorator_1.Destroy;
exports.scope = decorator_1.Scope;
exports.autowire = decorator_1.Autowire;
exports.priority = decorator_1.Priority;
exports.schedule = decorator_1.Schedule;
exports.config = decorator_1.Config;
exports.logger = decorator_1.Logger;
exports.plugin = decorator_1.Plugin;
exports.controller = decorator_1.Controller;
exports.session = decorator_1.Session;
exports.body = decorator_1.Body;
exports.query = decorator_1.Query;
exports.param = decorator_1.Param;
exports.headers = decorator_1.Headers;
exports.file = decorator_1.File;
exports.files = decorator_1.Files;
exports.post = decorator_1.Post;
exports.get = decorator_1.Get;
exports.del = decorator_1.Del;
exports.put = decorator_1.Put;
exports.patch = decorator_1.Patch;
exports.options = decorator_1.Options;
exports.head = decorator_1.Head;
exports.all = decorator_1.All;
exports.app = decorator_1.App;
exports.ScopeEnum = decorator_1.ScopeEnum;
var loader_1 = require("./loader/loader");
exports.AgentWorkerLoader = loader_1.AgentWorkerLoader;
exports.AppWorkerLoader = loader_1.AppWorkerLoader;
var midway_1 = require("./midway");
exports.Application = midway_1.Application;
exports.Agent = midway_1.Agent;
var webLoader_1 = require("./loader/webLoader");
exports.MidwayWebLoader = webLoader_1.MidwayWebLoader;
__export(require("@midwayjs/core"));
var egg_1 = require("egg");
exports.EggApplication = egg_1.EggApplication;
exports.Service = egg_1.Service;
exports.Boot = egg_1.Boot;
var egg_logger_1 = require("egg-logger");
exports.EggLogger = egg_logger_1.EggLogger;
exports.EggLoggers = egg_logger_1.EggLoggers;
exports.EggContextLogger = egg_logger_1.EggContextLogger;
