// midway/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EggContextLogger = exports.EggLoggers = exports.EggLogger = exports.Boot = exports.Service = exports.EggApplication = exports.ScopeEnum = exports.all = exports.head = exports.options = exports.patch = exports.put = exports.del = exports.get = exports.post = exports.files = exports.file = exports.headers = exports.param = exports.query = exports.body = exports.session = exports.controller = exports.plugin = exports.logger = exports.config = exports.schedule = exports.priority = exports.autowire = exports.scope = exports.destroy = exports.init = exports.async = exports.inject = exports.provide = exports.providerWrapper = exports.RELEASE = exports.VERSION = void 0;
const core_1 = require("@midwayjs/core");
/**
 * current version of midway
 * @member {String} Midway#VERSION
 */
exports.VERSION = require('../package.json').version;
/**
 * current release name
 * @member {String} Midway#RELEASE
 */
exports.RELEASE = 'WANDA';
/**
 * @deprecated Please import from @midwayjs/core
 */
exports.providerWrapper = core_1.providerWrapper;
/**
 * @deprecated Please import from @midwayjs/decorator
 */
var decorator_1 = require("@midwayjs/decorator");
Object.defineProperty(exports, "provide", { enumerable: true, get: function () { return decorator_1.Provide; } });
Object.defineProperty(exports, "inject", { enumerable: true, get: function () { return decorator_1.Inject; } });
Object.defineProperty(exports, "async", { enumerable: true, get: function () { return decorator_1.Async; } });
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return decorator_1.Init; } });
Object.defineProperty(exports, "destroy", { enumerable: true, get: function () { return decorator_1.Destroy; } });
Object.defineProperty(exports, "scope", { enumerable: true, get: function () { return decorator_1.Scope; } });
Object.defineProperty(exports, "autowire", { enumerable: true, get: function () { return decorator_1.Autowire; } });
Object.defineProperty(exports, "priority", { enumerable: true, get: function () { return decorator_1.Priority; } });
Object.defineProperty(exports, "schedule", { enumerable: true, get: function () { return decorator_1.Schedule; } });
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return decorator_1.Config; } });
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return decorator_1.Logger; } });
Object.defineProperty(exports, "plugin", { enumerable: true, get: function () { return decorator_1.Plugin; } });
Object.defineProperty(exports, "controller", { enumerable: true, get: function () { return decorator_1.Controller; } });
Object.defineProperty(exports, "session", { enumerable: true, get: function () { return decorator_1.Session; } });
Object.defineProperty(exports, "body", { enumerable: true, get: function () { return decorator_1.Body; } });
Object.defineProperty(exports, "query", { enumerable: true, get: function () { return decorator_1.Query; } });
Object.defineProperty(exports, "param", { enumerable: true, get: function () { return decorator_1.Param; } });
Object.defineProperty(exports, "headers", { enumerable: true, get: function () { return decorator_1.Headers; } });
Object.defineProperty(exports, "file", { enumerable: true, get: function () { return decorator_1.File; } });
Object.defineProperty(exports, "files", { enumerable: true, get: function () { return decorator_1.Files; } });
Object.defineProperty(exports, "post", { enumerable: true, get: function () { return decorator_1.Post; } });
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return decorator_1.Get; } });
Object.defineProperty(exports, "del", { enumerable: true, get: function () { return decorator_1.Del; } });
Object.defineProperty(exports, "put", { enumerable: true, get: function () { return decorator_1.Put; } });
Object.defineProperty(exports, "patch", { enumerable: true, get: function () { return decorator_1.Patch; } });
Object.defineProperty(exports, "options", { enumerable: true, get: function () { return decorator_1.Options; } });
Object.defineProperty(exports, "head", { enumerable: true, get: function () { return decorator_1.Head; } });
Object.defineProperty(exports, "all", { enumerable: true, get: function () { return decorator_1.All; } });
Object.defineProperty(exports, "ScopeEnum", { enumerable: true, get: function () { return decorator_1.ScopeEnum; } });
/**
 * @deprecated
 */
var egg_1 = require("egg");
Object.defineProperty(exports, "EggApplication", { enumerable: true, get: function () { return egg_1.EggApplication; } });
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return egg_1.Service; } });
Object.defineProperty(exports, "Boot", { enumerable: true, get: function () { return egg_1.Boot; } });
/**
 * @deprecated
 */
var egg_logger_1 = require("egg-logger");
Object.defineProperty(exports, "EggLogger", { enumerable: true, get: function () { return egg_logger_1.EggLogger; } });
Object.defineProperty(exports, "EggLoggers", { enumerable: true, get: function () { return egg_logger_1.EggLoggers; } });
Object.defineProperty(exports, "EggContextLogger", { enumerable: true, get: function () { return egg_logger_1.EggContextLogger; } });
__exportStar(require("@midwayjs/web"), exports);
