// @midwayjs/core/dist/service/informationService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MidwayInformationService = void 0;
const util_1 = require("../util");
const path_1 = require("path");
class MidwayInformationService {
    constructor(options) {
        this.baseDir = options.baseDir;
        this.appDir = options.appDir;
        if (!this.appDir) {
            this.appDir = (0, path_1.dirname)(this.baseDir);
        }
        this.pkg = (0, util_1.safeRequire)((0, path_1.join)(this.appDir, 'package.json')) || {};
    }
    getAppDir() {
        return this.appDir;
    }
    getBaseDir() {
        return this.baseDir;
    }
    getHome() {
        return (0, util_1.getUserHome)();
    }
    getPkg() {
        return this.pkg;
    }
    getProjectName() {
        var _a;
        return ((_a = this.pkg) === null || _a === void 0 ? void 0 : _a['name']) || '';
    }
    getRoot() {
        const isDevelopmentEnv = (0, util_1.isDevelopmentEnvironment)((0, util_1.getCurrentEnvironment)());
        return isDevelopmentEnv ? this.getAppDir() : this.getHome();
    }
}
exports.MidwayInformationService = MidwayInformationService;
