// @midwayjs/decorator/dist/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerlessTriggerType = exports.MidwayFrameworkType = exports.MSListenerType = exports.MSProviderType = void 0;
var MSProviderType;
(function (MSProviderType) {
    MSProviderType["DUBBO"] = "dubbo";
    MSProviderType["GRPC"] = "gRPC";
    MSProviderType["HSF"] = "hsf";
})(MSProviderType = exports.MSProviderType || (exports.MSProviderType = {}));
var MSListenerType;
(function (MSListenerType) {
    MSListenerType["RABBITMQ"] = "rabbitmq";
    MSListenerType["MQTT"] = "mqtt";
    MSListenerType["KAFKA"] = "kafka";
    MSListenerType["REDIS"] = "redis";
})(MSListenerType = exports.MSListenerType || (exports.MSListenerType = {}));
var MidwayFrameworkType;
(function (MidwayFrameworkType) {
    MidwayFrameworkType["WEB"] = "@midwayjs/web";
    MidwayFrameworkType["WEB_KOA"] = "@midwayjs/koa";
    MidwayFrameworkType["WEB_EXPRESS"] = "@midwayjs/express";
    MidwayFrameworkType["FAAS"] = "@midwayjs/faas";
    MidwayFrameworkType["MS_HSF"] = "";
    MidwayFrameworkType["MS_GRPC"] = "@midwayjs/grpc";
    MidwayFrameworkType["MS_RABBITMQ"] = "@midwayjs/rabbitmq";
    MidwayFrameworkType["WS_IO"] = "@midwayjs/socketio";
    MidwayFrameworkType["WS"] = "@midwayjs/ws";
    MidwayFrameworkType["SERVERLESS_APP"] = "@midwayjs/serverless-app";
    MidwayFrameworkType["CUSTOM"] = "";
    MidwayFrameworkType["EMPTY"] = "empty";
    MidwayFrameworkType["LIGHT"] = "light";
})(MidwayFrameworkType = exports.MidwayFrameworkType || (exports.MidwayFrameworkType = {}));
var ServerlessTriggerType;
(function (ServerlessTriggerType) {
    ServerlessTriggerType["EVENT"] = "event";
    ServerlessTriggerType["HTTP"] = "http";
    ServerlessTriggerType["API_GATEWAY"] = "apigw";
    ServerlessTriggerType["OS"] = "os";
    ServerlessTriggerType["CDN"] = "cdn";
    ServerlessTriggerType["LOG"] = "log";
    ServerlessTriggerType["TIMER"] = "timer";
    ServerlessTriggerType["MQ"] = "mq";
    ServerlessTriggerType["HSF"] = "hsf";
    ServerlessTriggerType["MTOP"] = "mtop";
})(ServerlessTriggerType = exports.ServerlessTriggerType || (exports.ServerlessTriggerType = {}));
