// @midwayjs/web/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.startCluster = exports.Agent = exports.Application = exports.createAgentWorkerLoader = exports.createAppWorkerLoader = exports.createEggAgent = exports.createEggApplication = exports.MidwayWebFramework = exports.Framework = void 0;
__exportStar(require("./interface"), exports);
var singleProcess_1 = require("./framework/singleProcess");
Object.defineProperty(exports, "Framework", { enumerable: true, get: function () { return singleProcess_1.MidwayWebSingleProcessFramework; } });
var web_1 = require("./framework/web");
Object.defineProperty(exports, "MidwayWebFramework", { enumerable: true, get: function () { return web_1.MidwayWebFramework; } });
var base_1 = require("./base");
Object.defineProperty(exports, "createEggApplication", { enumerable: true, get: function () { return base_1.createEggApplication; } });
Object.defineProperty(exports, "createEggAgent", { enumerable: true, get: function () { return base_1.createEggAgent; } });
Object.defineProperty(exports, "createAppWorkerLoader", { enumerable: true, get: function () { return base_1.createAppWorkerLoader; } });
Object.defineProperty(exports, "createAgentWorkerLoader", { enumerable: true, get: function () { return base_1.createAgentWorkerLoader; } });
var application_1 = require("./application");
Object.defineProperty(exports, "Application", { enumerable: true, get: function () { return application_1.Application; } });
Object.defineProperty(exports, "Agent", { enumerable: true, get: function () { return application_1.Agent; } });
var egg_1 = require("egg");
Object.defineProperty(exports, "startCluster", { enumerable: true, get: function () { return egg_1.startCluster; } });
